pkg/skaffold/yamltags/tags.go

```diff
@@ -77,37 +75,51 @@ func GetYamlTag(value interface{}) string {
 	return rawStr[:i]
 }
 
-// GetYamlTags returns the tags of the non-nested fields of the given non-nil value
-// If value interface{} is
-// latest.DeployType{HelmDeploy: &HelmDeploy{...}, KustomizeDeploy: &KustomizeDeploy{...}}
-// then this parses that interface as it's raw yaml:
-// 	kubectl:
-//    manifests:
-//    - k8s/*.yaml
-//  kustomize:
-//    paths:
-//    - k8s/
-// and returns ["kubectl", "kustomize"]
-// empty structs (e.g. latest.DeployType{}) when parsed look like "{}"" and so this function returns []
-func GetYamlTags(value interface{}) []string {
+// GetYamlKeys returns the yaml key for each non-nested field of the given non-nil config parameter
+// For example if config is `latest.DeployType{HelmDeploy: &HelmDeploy{...}, KustomizeDeploy: &KustomizeDeploy{...}}`
+// then it returns `["helm", "kustomize"]`
+func GetYamlKeys(config interface{}) []string {
 	var tags []string
-
-	buf, err := yaml.Marshal(value)
-	if err != nil {
-		logrus.Warnf("error marshaling %-v", value)
+	if config == nil {
 		return tags
 	}
+	st := reflect.Indirect(reflect.ValueOf(config))
+	t := st.Type()
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		v := st.Field(i)
+		if v.Kind() == reflect.Ptr && v.IsNil() { // exclude ptr fields not explicitly defined in the configuration
+			continue
+		}
+		tag := getYamlKey(f)
+		if tag != "" {
+			tags = append(tags, tag)
+		}
+	}
+	return tags
+}
 
-	r := bufio.NewScanner(bytes.NewBuffer(buf))
-	for r.Scan() {
-		l := r.Text()
-		i := strings.Index(l, ":")
-		if !unicode.IsSpace(rune(l[0])) && l[0] != '-' && i >= 0 {
-			tags = append(tags, l[:i])
+func getYamlKey(f reflect.StructField) string {
+	t, ok := f.Tag.Lookup("yaml")
```

Comment:
I suppose you should also check whether it's exported (starts with a capital letter)?